components/VM_Arm/src/fdt_manipulation.c

```diff
@@ -47,59 +47,136 @@ int fdt_generate_memory_node(void *fdt, unsigned long base, size_t size)
     return 0;
 }
 
+int fdt_generate_sel4_tracebuffer_node(void *fdt, unsigned long base, size_t size)
+{
+    int root_offset = fdt_path_offset(fdt, "/reserved-memory");
+    int address_cells = fdt_address_cells(fdt, root_offset);
+    int size_cells = fdt_size_cells(fdt, root_offset);
+
+    ZF_LOGE("/reserved-memory node: 0x%x", root_offset);
+
+    int this = fdt_add_subnode(fdt, root_offset, "sel4_tracebuffer");
+    if (this < 0) {
+        ZF_LOGE("Can not create sel4_tracebuffer node");
+        return this;
+    }
+
+    int err = fdt_appendprop_string(fdt, this, "compatible", "sel4_tracebuffer");
+    if (err) {
+        ZF_LOGE("Can not set compatible prop");
+        return err;
+    }
+
+    err = append_prop_with_cells(fdt, this, base, address_cells, "reg");
+    if (err) {
+        ZF_LOGE("Can not add reg prop");
+        return err;
+    }
+
+    err = append_prop_with_cells(fdt, this, size, size_cells, "reg");
+    if (err) {
+        ZF_LOGE("Can not add reg2 prop");
+        return err;
+    }
+
+    return 0;
+}
+
+int fdt_generate_ramoops_node(void *fdt, unsigned long base, size_t size)
+{
+    int root_offset = fdt_path_offset(fdt, "/reserved-memory");
+    int address_cells = fdt_address_cells(fdt, root_offset);
+    int size_cells = fdt_size_cells(fdt, root_offset);
+
+    ZF_LOGE("/reserved-memory node: 0x%x", root_offset);
+
+    int this = fdt_add_subnode(fdt, root_offset, "ramoops");
+    if (this < 0) {
+        ZF_LOGE("Can not create ramoops node");
+        return this;
+    }
+
+    int err = fdt_appendprop_string(fdt, this, "compatible", "ramoops");
+    if (err) {
+        ZF_LOGE("Can not set compatible prop");
+        return err;
+    }
+
+    err = append_prop_with_cells(fdt, this, base, address_cells, "reg");
+    if (err) {
+        ZF_LOGE("Can not add reg prop");
+        return err;
+    }
+
+    err = append_prop_with_cells(fdt, this, size, size_cells, "reg");
+    if (err) {
+        ZF_LOGE("Can not add reg2 prop");
+        return err;
+    }
+
+    err = append_prop_with_cells(fdt, this, size, 1, "ftrace-size");
+    if (err) {
+        ZF_LOGE("Can not add ftrace-size");
+        return err;
+    }
+
+    return 0;
+}
+
 int fdt_generate_usb_node(void *fdt)
 {
     int root_offset = fdt_path_offset(fdt, "/scb/pcie@7d500000/pci@0,0");
     int address_cells = fdt_address_cells(fdt, root_offset);
     int size_cells = fdt_size_cells(fdt, root_offset);
 
-    if (root_offset <= 0)
-       ZF_LOGE("root_offset: %d", root_offset);
+    if (root_offset <= 0) {
```

Comment:
Hmmm while these whitespace and similar style fixes are definitely good, can you put them into separate commit? That commit can be part of this PR, there's no problem. And the commit message could be something like "trivial: Fix whitespace".

Reply:
Done.